### PR TITLE
Dom api task

### PR DESCRIPTION
**Pull request by Ant-C-tech**
### task_js_dom

[Demo](https://ant-c-tech.github.io/-7-Document-Object-Model-Practice/) |
[Code base](https://github.com/Ant-C-tech/-7-Document-Object-Model-Practice)

The code is submitted in a dedicated feature branch.
Please, review.

### CODE REVIEW

File: submissions/Ant-C-tech/dom-api-task/js/app.js [hunk 1]
```diff
@@ -0,0 +1,881 @@
...
+        })
+        //Remove hint on mouseleave, hover
+        iterator.addEventListener('mouseleave', function (event) {
+            if (event.target.style.fill !== 'rgb(0, 176, 255)') {
+                event.target.style.fill = '#000'
+            }
+            hintSection.innerHTML = '. . .'
+        })
+        //Change content on click
+        iterator.addEventListener('click', function (event) {
```
Comment (babayK0) [inline submissions/Ant-C-tech/dom-api-task/js/app.js:765]: I think one listener (for clicks) for `map` will be better.

Comment (babayK0) [inline submissions/Ant-C-tech/dom-api-task/js/app.js:765]: you can use object destructuring 
```
({ target }) {
 ...
}
```

File: submissions/Ant-C-tech/dom-api-task/js/app.js [hunk 2]
```diff
@@ -0,0 +1,881 @@
...
+    contentSection.classList.add('content-hide')
+}
+
+function changeContent(event) {
+    hideContent()
+    for (let item of mapAreas) {
+        item.style.fill = '#000'
+    }
+    event.target.style.fill = '#00b0ff'
+    const target = getTargetId(event)
```
Comment (babayK0) [inline submissions/Ant-C-tech/dom-api-task/js/app.js:816]: Change naming. This one is not target. It is ID.

You can get id with destructuring.

`const { id } = event.target`  or `const { id } = target` (if you can access `target` directly  )

File: submissions/Ant-C-tech/dom-api-task/js/app.js [hunk 3]
```diff
@@ -0,0 +1,881 @@
...
+    contentSection.addEventListener('transitionend', function () {
+        contentSection.innerHTML = ''
+        showContent(content[target])
+    }, { once: true })
+
+}
+
+function getTargetId(event) {
+    return event.target.id
+}
```
Comment (babayK0) [inline submissions/Ant-C-tech/dom-api-task/js/app.js:826]: You dont need this function

**Review by babayK0 — COMMENTED** [review]
Wow, such a beautiful work :heart_eyes:

Let improve your code together! 

Firstly, there are many comments that you do not need. Don't write comments for everything, especially on obvious things.  Clean and well-structured code is self-documented.

Secondly, when i click and content changes your `flexWrapper` shrinks. Maybe add  `min-height`  for this block ? ( or you can suggest other implementation)

When i click in `Ukraine` block multiple times it causes content changing. Add check for double click. 

Watch my comments below and let start first iteration of improving :wink:

File: submissions/Ant-C-tech/dom-api-task/js/app.js [hunk 4]
```diff
@@ -0,0 +1,881 @@
...
+function hideContent() {
+    contentSection.classList.add('content-hide')
+}
+
+function changeContent(event) {
+    hideContent()
+    for (let item of mapAreas) {
+        item.style.fill = '#000'
+    }
+    event.target.style.fill = '#00b0ff'
```
Comment (babayK0) [inline submissions/Ant-C-tech/dom-api-task/js/app.js:815]: Use css

**Review by babayK0 — CHANGES_REQUESTED** [review]

**Comment by Ant-C-tech** [comment]
First, I want to say thank you for your review. It made me thought again about many aspects of my code. I formerly worked as an engineer and I good undestanding how it is boring to check students works. But for me I think it was the most important part of this course. Now I understand why nobody want to work with people without experience. You made greate job. 
I tried to made revisions step by step, commit by commit, but it is hard to me yet, because I planed my work bad.  
I found a problem in my app with fast changing of content, so it  made me changed the algoritm of changing content. I hope it will not make your work more difficult.
Regards.

File: submissions/Ant-C-tech/dom-api-task/js/app.js [hunk 5]
```diff
@@ -808,21 +77,30 @@ function hideContent() {
...
-    return event.target.id
+    if (event.target.classList.contains('land-active')) {
+        return false
+    } else {
+        //Map visualization
+        for (const land of mapAreas) {
+            land.classList.remove('land-active')
+        }
+        map.removeEventListener('click', changeContent)
+        event.target.classList.toggle('land-active')
```
Comment (babayK0) [inline submissions/Ant-C-tech/dom-api-task/js/app.js:90]: `classList.add` i think will be more understandable in this case. Do the same, but you don't need any `toggling` here. Previously you removed all active classes, so main feature of toggle will not used

File: submissions/Ant-C-tech/dom-api-task/js/app.js [hunk 6]
```diff
@@ -797,6 +62,10 @@ function recalcMap() {
     }
 }
 
+function recalcContentBlock() {
+    contentSection.style.minHeight = mapContainer.offsetHeight + 'px'
+}
+
 function showContent(param) {
```
Comment (babayK0) [inline submissions/Ant-C-tech/dom-api-task/js/app.js:69]: change naming. `param` is not obvious. Maybe id will be better ?

File: submissions/Ant-C-tech/dom-api-task/js/app.js [hunk 7]
```diff
@@ -808,21 +77,30 @@ function hideContent() {
...
+        map.removeEventListener('click', changeContent)
+        event.target.classList.toggle('land-active')
+
+        //Change content
+        country.removeEventListener('click', changeContent)
+        hideContent()
+        contentSection.addEventListener('transitionend', function () {
+            contentSection.innerHTML = ''
+            showContent(content[event.target.id])
+            contentSection.addEventListener('transitionend', function () {
```
Comment (babayK0) [inline submissions/Ant-C-tech/dom-api-task/js/app.js:98]: Why here you have nested 'transitionend' listener ?

**Review by babayK0 — CHANGES_REQUESTED** [review]

File: submissions/Ant-C-tech/dom-api-task/js/app.js [hunk 8]
```diff
@@ -0,0 +1,157 @@
...
+        img.setAttribute('src', `${item.src}`)
+        img.setAttribute('alt', `${item.descript}`)
+        figure.appendChild(img)
+        let figcaption = document.createElement('figcaption')
+        figcaption.classList.add('imgBlock__caption')
+        figcaption.textContent = item.descript
+        figure.appendChild(figcaption)
+        imgBlock.appendChild(figure)
+    }
+    fragment.appendChild(imgBlock)
```
Comment (babayK0) [inline submissions/Ant-C-tech/dom-api-task/js/app.js:148]: You can append many elements with one `append`

 [append](https://developer.mozilla.org/en-US/docs/Web/API/ParentNode/append)

**Comment by babayK0** [comment]
@Ant-C-tech 

> First, I want to say thank you for your review. It made me thought again about many aspects of my code. I formerly worked as an engineer and I good undestanding how it is boring to check students works. But for me I think it was the most important part of this course. Now I understand why nobody want to work with people without experience. You made greate job.
> I tried to made revisions step by step, commit by commit, but it is hard to me yet, because I planed my work bad.
> I found a problem in my app with fast changing of content, so it made me changed the algoritm of changing content. I hope it will not make your work more difficult.
> Regards.

Thanks for your feedback :smile_cat:

**Comment by Ant-C-tech** [comment]
Once again thank you for your review. Every time I read your notices I think how can I do not see these things. But I really can't yet. So it is very useful for me to work with your reviews.
With one review I did nothing so I want to explain my thoughts and maybe you can help me to make my code better.
It is about your review "Why here you have nested 'transitionend' listener ?". In this case I wanted to construct a chain of animation and events those have to occur consistently. First: turning off all click listeners, than  animation of hide content, than - change content outside the screen, than  - animation of showing content and than turning on all the click listeners. I guess I could make with function raf(fn) {
    window.requestAnimationFrame(function () {
        window.requestAnimationFrame(function () {
            fn()
        })
    })
} (it works not always properly - I think I do not understand it enough) or maybe promices (I do not understand them at all). So I could make my idea only with nested listeners.

**Comment by babayK0** [comment]
@Ant-C-tech 

> Once again thank you for your review. Every time I read your notices I think how can I do not see these things. But I really can't yet. 

Code review was invented to prevent "raw" on production. You cant to know everything. with help of other developers, resulting code on production will be improved, more clean, readable etc.

File: submissions/Ant-C-tech/dom-api-task/js/app.js [hunk 9]
```diff
@@ -0,0 +1,154 @@
...
+    }, { once: true })
+
+}
+
+function createContent(content) {
+    let fragment = new DocumentFragment()
+
+    let titleBlock = document.createElement('div')
+    let title = document.createElement('div')
+    let subtitle = document.createElement('div')
```
Comment (babayK0) [inline submissions/Ant-C-tech/dom-api-task/js/app.js:112]: Why not use const here ?

**Comment by Ant-C-tech** [comment]
Thank you for you review. It is interesting, I have never heard about pointer-event. Now I have a bit different behave of hover but I think now it is more logical.

**Comment by babayK0** [comment]
Perfect job !!!

> One more I have noticed, there was one problem - when mousedown event happens on the one map area
> and mouseup event happens on the other map area
> application was crushed. I think, I fixed it but I am not sure that I did it in the right way.

It is not blocker i guess, so i can approve you pr now, but if you want we can debug this bug together.
Can you provide a 'broken' version in another branch? (don't open new pr, just send me link to branch with bug)

**Review by babayK0 — APPROVED** [review]

**Pull request merged**